File: lib/js/src/Json_encode.js
'use strict';

var List       = require("bs-platform/lib/js/list.js");
var $$Array    = require("bs-platform/lib/js/array.js");
var Curry      = require("bs-platform/lib/js/curry.js");
var Js_dict    = require("bs-platform/lib/js/js_dict.js");
var Js_boolean = require("bs-platform/lib/js/js_boolean.js");

var bool = Js_boolean.to_js_boolean;

function date(d) {
  return d.toJSON();
}

function nullable(encode, param) {
  if (param) {
    return Curry._1(encode, param[0]);
  } else {
    return null;
  }
}

function withDefault(d, encode, param) {
  if (param) {
    return Curry._1(encode, param[0]);
  } else {
    return d;
  }
}

var object_ = Js_dict.fromList;

var arrayOf = $$Array.map;

function list(encode, l) {
  return $$Array.of_list(List.map(encode, l));
}

function pair(encodeA, encodeB, param) {
  return /* array */[
          Curry._1(encodeA, param[0]),
          Curry._1(encodeB, param[1])
        ];
}

function tuple3(encodeA, encodeB, encodeC, param) {
  return /* array */[
          Curry._1(encodeA, param[0]),
          Curry._1(encodeB, param[1]),
          Curry._1(encodeC, param[2])
        ];
}

function tuple4(encodeA, encodeB, encodeC, encodeD, param) {
  return /* array */[
          Curry._1(encodeA, param[0]),
          Curry._1(encodeB, param[1]),
          Curry._1(encodeC, param[2]),
          Curry._1(encodeD, param[3])
        ];
}

var tuple2 = pair;

exports.bool        = bool;
exports.date        = date;
exports.nullable    = nullable;
exports.withDefault = withDefault;
exports.pair        = pair;
exports.tuple2      = tuple2;
exports.tuple3      = tuple3;
exports.tuple4      = tuple4;
exports.object_     = object_;
exports.arrayOf     = arrayOf;
exports.list        = list;
/* Js_dict Not a pure module */

File: src/Json_encode.ml
type 'a encoder = 'a -> Js.Json.t

external null : Js.Json.t = "null" [@@bs.val]
external string : string -> Js.Json.t = "%identity"
external float : float -> Js.Json.t = "%identity"
external int : int -> Js.Json.t = "%identity"
external bool : bool -> Js.Json.t = "%identity"

let char c =
  c |> String.make 1
    |> string

let date d =
  d |> Js.Date.toJSONUnsafe
    |> string

let nullable encode = function
  | None -> null
  | Some v -> encode v

let withDefault d encode = function
  | None -> d
  | Some v -> encode v

external jsonDict : Js.Json.t Js_dict.t -> Js.Json.t = "%identity"
let dict encode d =
  let pairs = Js.Dict.entries d in
  let encodedPairs = Array.map (fun (k, v) -> (k, encode(v))) pairs in
  jsonDict (Js.Dict.fromArray encodedPairs)

let object_ props: Js.Json.t =
  props |> Js.Dict.fromList
        |> jsonDict

external jsonArray : Js.Json.t array -> Js.Json.t = "%identity"
let array encode l =
  l |> Array.map encode
    |> jsonArray
let list encode = function
  | [] -> jsonArray [||]
  | hd::tl as l ->
      let a = Array.make (List.length l) (encode hd) in
      let rec fill i = function
        | [] -> a
        | hd::tl -> (
          Array.unsafe_set a i (encode hd);
          fill (i + 1) tl
        )
      in
        jsonArray (fill 1 tl)

let pair encodeA encodeB (a, b) =
  jsonArray [|encodeA a; encodeB b|]
let tuple2 = pair
let tuple3 encodeA encodeB encodeC (a, b, c) =
  jsonArray [|encodeA a; encodeB b; encodeC c|]
let tuple4 encodeA encodeB encodeC encodeD (a, b, c, d) =
  jsonArray [|encodeA a; encodeB b; encodeC c; encodeD d|]

external stringArray : string array -> Js.Json.t = "%identity"
external numberArray : float array -> Js.Json.t = "%identity"
external boolArray : bool array -> Js.Json.t = "%identity"
